### DISCUSSION TITLE
HTTP filters or custom REST endpoint are not hit when non-application endpoints?

### DISCUSSION THREAD

**Discussion by ppatierno** [quarkusio/quarkus#32671] [Q&A]
Hi,
trying to "intercept" the HTTP requests on the non-application `/q/metrics` endpoint but it seems that any `@ServerResponseFilter` or `@ServerRequestFilter` method is just skipped.
Is this on purpose?

**Comment by ppatierno** [comment]
Actually I decided to not using a filter but defining a REST endpoint for it. So the question now is ... is it possible to "override" the `/q/metrics` endpoint in order to have my code to be executed instead of the Quarkus built-in one?

**Reply by ebullient** [reply]
If you use the micrometer- extension, you get the default handler. The default handler calls the scrape method on whatever PrometheusMeterRegistry is registered:
https://github.com/quarkusio/quarkus/discussions/32647#discussioncomment-5636668

The alterative is to create your own MeterRegistry, and ditch the prometheus one, which is easier than it sounds. ;) Plenty of examples here: 
https://github.com/quarkiverse/quarkus-micrometer-registry

You'd be specifically interested in registering your own non-application endpoint (recommend dropping to vertx over servlets), you'd want to follow what the prometheus processor (crossed with other examples from the quarkiverse repo) is doing here: 

https://github.com/quarkusio/quarkus/blob/b687fbb7ba8678d99433be71aca5dd1f358fb6e6/extensions/micrometer/deployment/src/main/java/io/quarkus/micrometer/deployment/export/PrometheusRegistryProcessor.java

**Reply by ppatierno** [reply]
Thanks @ebullient I think what's going on in the other discussion will meet our requirements.

**Reply by ppatierno** [reply]
@ebullient sorry to re-open this but ...

@ShubhamRwt implemented our "custom" registry the way you suggested here ...
https://github.com/ppatierno/strimzi-kafka-bridge/commit/7888df30420318d9d170a2bc088df66fc07d72fc

and also we are now exposing the `/q/metrics` as `/metrics` (to take backward compatibility for our users), by using:
`quarkus.micrometer.export.prometheus.path=/metrics`

The problem is that we have this `RestLoggingFilter` which should be used for logging purposes but it doesn't kick in when we hit the `/metrics` (or `/q/metrics` endpoint).
https://github.com/ppatierno/strimzi-kafka-bridge/blob/road-to-quarkus/src/main/java/io/strimzi/kafka/bridge/quarkus/RestLoggingFilter.java

I can understand it's a non-application endpoint and maybe you want to avoid filter code doing something wrong.
What I can think of is creating a REST method in our API for the `/metrics` and calling the scrape on the Prometheus registry instance directly. Maybe you suggested that in the other discussion but we didn't get right?

While this should work, we should disable the `/q/metrics` endpoint. Is it possible?

**Reply by ebullient** [reply]
Non-application endpoints are not rest endpoints, they are handled by vertx directly. Rest filters will not be invoked on that path.

Are you trying to have all requests logged? have you tried using the Quarkus http access log? https://quarkus.io/guides/http-reference#configuring-http-access-logs

**Reply by ppatierno** [reply]
It's not so simple for us ... due to the Quarkus logging system we have already broken some backward compatibility on our Strimzi HTTP bridge for the users, which is pretty bad. 
I don't see a way to use the HTTP access logs, unless users can configure to select a specific logging level for each REST endpoint (OpenAPI operation) this way:

```
quarkus.log.category."http.openapi.operation.healthy".level=WARN
quarkus.log.category."http.openapi.operation.ready".level=DEBUG
```

which already broke how it is done today with the following in a log4j2.properties file:

```
# HTTP OpenAPI specific logging levels (default is INFO)
# Logging healthy and ready endpoints is very verbose because of Kubernetes health checking.
logger.healthy.name = http.openapi.operation.healthy
logger.healthy.level = WARN
logger.ready.name = http.openapi.operation.ready
logger.ready.level = WARN
```

More details if you are interested in this discussion https://github.com/quarkusio/quarkus/discussions/32078

So I think a warkaround for us is declaring our own `/metrics` REST endpoint in order to scrape metrics but we need to disable `/q/metrics` to avoid having a duplication. Is it possible?

**Reply by ebullient** [reply]
You are so close now.. you can just remove the prometheus micrometer extension (and just include the base extension)..

You already have a meter registry producer (that will get picked up by the core micrometer extension)...

**Comment by ShubhamRwt** [comment]
We are currently using 
```
		<dependency>
			<groupId>io.quarkus</groupId>
			<artifactId>quarkus-micrometer-registry-prometheus</artifactId>
		</dependency>
```
so ,what you are suggesting is that we can use this instead
```
		<dependency>
			<groupId>io.quarkus</groupId>
			<artifactId>quarkus-micrometer</artifactId>
		</dependency>
``` 
right? If yes, then the other problem is that we are are extending `PrometheusMeterRegistry` which is coming through the Prometheus extension

**Reply by ShubhamRwt** [reply]
@ebullient ^^

**Reply by ebullient** [reply]
Correct. You need to bring in the prometheus registry explicitly:
```
<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-micrometer</artifactId>
</dependency>
<dependency>
    <groupId>io.micrometer</groupId>
    <artifactId>micrometer-registry-prometheus</artifactId>
</dependency>
```

You are already producing a `PrometheusMeterRegistry`. You may need to include this:

https://github.com/quarkusio/quarkus/blob/main/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/export/PrometheusMeterRegistryProvider.java

**Reply by ShubhamRwt** [reply]
@ebullient I have done this step but I think that the `\q\metrics` endpoint is still enabled? Can't we disable it somehow?

**Reply by ppatierno** [reply]
@ebullient it's not clear how this would help to have our HTTP filter to kick in when the `/metrics` or `/q/metrics` endpoint is hit.
We tried but didn't work.
Or we are overlapping two different discussions?

**Reply by ShubhamRwt** [reply]
@ebullient Hi, Any suggestions/workarounds regarding this ?
